### ISSUE TITLE
Consider merging GPUSwapChain into GPUCanvasContext

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#1780]
_Originally posted by @toji in https://github.com/gpuweb/gpuweb/issues/1691#issuecomment-849903077_

> IMO that would be okay. It's a simple state check, and for an obvious reason (no GPUDevice has been provided yet).

If we do want to collapse swap chains and the context into one, I'd suggest retaining the `GPUSwapChain` name as the thing that `canvas.getContext()` returns. (Or `GPUSwapChainContext` if you really want to have the word "context" in there.) I really like having an interface that's a clear parallel to the native swap chain concept. As a bonus, you could make the context ID `gpuswapchain`, since dislike for `gpupresent` keeps coming up. 😁

Sketching it out real quick, it could look like this. Biggest questions become where errors get surfaced since the device that the swap chain is associated with is mutable and may be `null`.

```webidl
dictionary GPUSwapChainCreationArgs {
  GPUSwapChainDescriptor initialConfiguration;
}

interface GPUSwapChainContext {
  GPUTextureFormat getPreferredFormat(GPUAdapter adapter);

  // Need a way to surface errors here?
  undefined reconfigure(GPUSwapChainDescriptor descriptor);

  GPUTexture getCurrentTexture();
};
```

Usage then looks something like this:

```js
const swapChain = canvas.getContext('gpuswapchain');

// Would Error: Not configured yet.
//let texture = swapChain.getCurrentTexture();

swapChain.reconfigure({
  device,
  format: swapChain.getPreferredFormat(device.adpter)
  // size still defaults to canvas width/height?
});

// Now this works.
let texture = swapChain.getCurrentTexture();
```

Alternately, to allow it to be used immediately upon creation, you can do the following
BUT you don't have any way of querying the preferred format first, so you just have to go with a known good format.

```js
const swapChain = canvas.getContext('gpuswapchain', {
  initialConfiguration: {
    device,
    format: 'bgra8unorm'
  }
});

// Works immediately after creation.
let texture = swapChain.getCurrentTexture();
```

And of course I'd prefer that we keep the previously discussed property of the swap chain not resizing unless explicitly reconfigured.

```js
window.addEventListener('resize', () => {
  // Must explicitly reconfigure swap chain if you want to keep it in sync with
  // canvas resizes.
  swapChain.reconfigure({
    device,
    format: swapChain.getPreferredFormat(device.adpter),
    size: [canvas.offsetWidth * devicePixelRatio, canvas.offsetHeight * devicePixelRatio]
  });
});
```

(This is the reason I used `reconfigure` rather than `configure` as the method name, because it seems like the majority of the time it'll be changing the configuration rather than establishing an initial one.)

**Comment by Kangz** [comment]
+1 on the separation for `GPUCanvasContent` and `GPUSwapChain` being unnecessary if other APIs like WebXR aren't going to reuse `GPUSwapChain`. However there's still multiple questions to address:

 - Do we keep the name swapchain? Some folks have complained in the past that this object didn't really act like a swapchain in native, and that the name doesn't make too much sense for the Web. (I don't have an opinion on this).
 - Do you need to respcify the device (resp. the format) every time?
 - Is there a way to "detach" the device from the canvas context so that the device can be garbage collected while the <canvas> stays on the page?
 - Why `initialConfiguration` and not just passing in a dictionary that extends `GPUSwapChainDescriptor`?
 - What about the texture usages?
 - Is the change worth it this late? (imho probably, and it's not too intrusive)

**Comment by kvark** [comment]
Generally agree with this simplifications, and want to add a few concerns to @Kangz 's list:
  - the `getCurrentTexture` always returns a texture owned by the device that the context was last configured with. This means the device is an implicit state of the swapchain. Imagine the following code: there are 2 different pieces of logic, both get an `HTMLCanvasContext` on the input. Both of them configure the swapchain and then start rendering into it. What happens if they both get the same canvas context? One of these actors would get a texture belonging to an entirely different device. This is totally unexpected by them.
  - once the canvas context is configured with a device, what is the lifetime relationship here? Would the canvas hold the device alive while it's configured? Since there is no `unconfigure`, this seems a bit unfortunate.
  - why is `initialConfiguration` needed at all? They can just call `reconfigure` already.

**Comment by toji** [comment]
From editors sync:

 - No need for `initialConfiguration`.
 - In order to allow the device to be GCed we should allow the configuration to be cleared.
 - While the swap chain is not configured or the configuration has been explicitly cleared `getCurrentTexture()` can throw an exception.
 - While the swap chain is configured any errors are surfaced through the configured device.
 - Respecifying the device every time is fine, because there's no reason that the context should be bound to a single device for it's lifetime. (Plus there's no way to destroy a canvas context, you just have to let the owning canvas and the context get GCed).

With the above in mind, an updated proposal would look like this:

```webidl
interface GPUSwapChainContext {
  GPUTextureFormat getPreferredFormat(GPUAdapter adapter);
  
  // Note that the descriptor is nullable. Reconfiguring with `null` clears the configuration.
  undefined reconfigure(GPUSwapChainDescriptor? descriptor);
  
  // Throws an exception when the swap chain is not configured.
  GPUTexture getCurrentTexture();
};

// No context creation args.
```

Example usage:

```js
const swapChain = canvas.getContext('gpuswapchain');
const format = swapChain.getPreferredFormat(device.adapter);

try() {
  let badTexture = swapChain.getCurrentTexture();
} catch(error) {
  console.log("Error expected. Swap chain not yet configured.");
}

swapChain.reconfigure({
  device,
  format,
  // size still defaults to canvas width/height
});

// Now it's been configured, no longer throws.
let texture = swapChain.getCurrentTexture();

// Relies on typical dictionary validation to throw a type error. Does not change previously valid configuration.
swapChain.reconfigure({
  device: 5,
  format
});

// Still valid, since previous reconfigure failed.
let texture = swapChain.getCurrentTexture();

// reconfigure succeeds but configuration is bad. Reports an error on the current device error scope.
swapChain.reconfigure({
  device,
  format,
  size: [Number.MAX_SAFE_INTEGER, Number.MAX_SAFE_INTEGER]
});

// Returns an invalid texture, handled identically to if you requested a too-large texture from createTexture.
texture = swapChain.getCurrentTexture();

window.addEventListener('resize', () => {
  // Must still explicitly reconfigure swap chain if you want to keep it in sync with canvas resizes.
  swapChain.reconfigure({
    device,
    format,
    size: [canvas.offsetWidth * devicePixelRatio, canvas.offsetHeight * devicePixelRatio]
  });
});

// ...some time later, if the device needs to be destroyed/GCed:
swapChain.reconfigure(null); // Clears configuration

try() {
  let badTexture = swapChain.getCurrentTexture();
} catch(error) {
  console.log("Error expected. Swap chain is no longer configured.");
}
```

**Comment by kainino0x** [comment]
> * Is there a way to "detach" the device from the canvas context so that the device can be garbage collected while the  stays on the page?

My alternative proposal was that the canvas context could hold a weakref to the device, and vend "invalid" textures if the weakref is dead. On the one hand, this adds some implementation complexity (how do you vend an invalid texture without having a device at all? they would really have to be more like "empty shell" GPUTexture objects). On the other hand, if users are required to unconfigure the swapchain to release the device... I think they're unlikely to adhere to this consistently.

**Comment by kainino0x** [comment]
> * the `getCurrentTexture` always returns a texture owned by the device that the context was last configured with. This means the device is an implicit state of the swapchain. Imagine the following code: there are 2 different pieces of logic, both get an `HTMLCanvasContext` on the input. Both of them configure the swapchain and then start rendering into it. What happens if they both get the same canvas context? One of these actors would get a texture belonging to an entirely different device. This is totally unexpected by them.

I don't think there is a solution to this problem. We already have an equivalent problem today with swap chain objects.

**Comment by kainino0x** [comment]
> On the other hand, if users are required to unconfigure the swapchain to release the device... I think they're unlikely to adhere to this consistently.

Oh, and we could just as well say that they should just call `device.destroy()`.

**Comment by Kangz** [comment]
> Oh, and we could just as well say that they should just call device.destroy().

Seems like a good idea.

> `   undefined reconfigure(GPUSwapChainDescriptor? descriptor);`

nit: maybe this should be `configure` since it is also used for the first configuration.

**Comment by kvark** [comment]
@kainino0x 
> I don't think there is a solution to this problem. We already have an equivalent problem today with swap chain objects.

I don't think we have this problem today. You get the current texture from a swap chain, and it has a concrete device as a parent, which never changes behind your back.

**Comment by kainino0x** [comment]
The current problem is not exactly the same, but I think it's equivalent: instead of getting a texture with the wrong parent device, you'd end up with one of your swapchains being dead, which will error in some way when you try to getCurrentTexture.

**Comment by kvark** [comment]
In one case you are getting a current texture that's internally NULL/error. This is very much in line with all the other error handling we have currently. But in the new case, you are getting a texture that is totally valid, just owned by a different device. Maybe it doesn't make a difference for the user (either way it's an error to use it), so we are good, but the difference is worth consideration.

**Comment by kainino0x** [comment]
Resolution: tentatively accepted, moving to concrete proposal in #1810
